python/pyspark/pandas/data_type_ops/categorical_ops.py

```diff
@@ -70,28 +70,58 @@ def astype(self, index_ops: IndexOpsLike, dtype: Union[str, type, Dtype]) -> Ind
             scol = map_scol.getItem(index_ops.spark.column)
         return index_ops._with_new_scol(scol).astype(dtype)
 
+    def eq(self, left: IndexOpsLike, right: Any) -> SeriesOrIndex:
+        return _compare(left, right, Column.__eq__, is_equality_comparison=True)
+
+    def ne(self, left: IndexOpsLike, right: Any) -> SeriesOrIndex:
+        return _compare(left, right, Column.__ne__, is_equality_comparison=True)
+
     def lt(self, left: IndexOpsLike, right: Any) -> SeriesOrIndex:
-        _non_equality_comparison_input_check(left, right)
-        return column_op(Column.__lt__)(left, right)
+        return _compare(left, right, Column.__lt__)
 
     def le(self, left: IndexOpsLike, right: Any) -> SeriesOrIndex:
-        _non_equality_comparison_input_check(left, right)
-        return column_op(Column.__le__)(left, right)
+        return _compare(left, right, Column.__le__)
 
     def gt(self, left: IndexOpsLike, right: Any) -> SeriesOrIndex:
-        _non_equality_comparison_input_check(left, right)
-        return column_op(Column.__gt__)(left, right)
+        return _compare(left, right, Column.__gt__)
 
     def ge(self, left: IndexOpsLike, right: Any) -> SeriesOrIndex:
-        _non_equality_comparison_input_check(left, right)
-        return column_op(Column.__ge__)(left, right)
+        return _compare(left, right, Column.__ge__)
 
 
-def _non_equality_comparison_input_check(left: IndexOpsLike, right: Any) -> None:
-    if not cast(CategoricalDtype, left.dtype).ordered:
-        raise TypeError("Unordered Categoricals can only compare equality or not.")
+def _compare(
+    left: IndexOpsLike,
+    right: Any,
+    f: Callable[..., Column],
+    *,
+    is_equality_comparison: bool = False
+) -> SeriesOrIndex:
+    """
+    Compare a Categorical operand `left` to `right` with the given Spark Column function.
+
+    Parameters
+    ----------
+    left: A Categorical operand
+    right: The other operand to compare with
+    f : The Spark Column function to apply
+    is_equality_comparison: True if it is equality comparison, ie. == or !=. False by default.
+
+    Returns
+    -------
+    SeriesOrIndex
+    """
     if isinstance(right, IndexOpsMixin) and isinstance(right.dtype, CategoricalDtype):
+        if not is_equality_comparison:
+            if not cast(CategoricalDtype, left.dtype).ordered:
+                raise TypeError("Unordered Categoricals can only compare equality or not.")
         if hash(left.dtype) != hash(right.dtype):
             raise TypeError("Categoricals can only be compared if 'categories' are the same.")
+        return column_op(f)(left, right)
+    elif not hasattr(right, "__len__"):
```

Comment:
This is not enough. It rejects `str` values.

```py
>>> hasattr('abc', '__len__')
True
```

Reply:
Could you add some tests with string categories as well?

Reply:
Makes sense. [`np.isscalar`](https://numpy.org/doc/stable/reference/generated/numpy.isscalar.html) is used instead.

Reply:
String categories are tested now.

Reply:
Can we use `isscalar` here? What if other scalar values, e.g., `datetime.date`, `datetime.datetime`, etc.?

Reply:
Hmm, you are right. I changed it to `isinstance(right, str) or not hasattr(right, "__len__")`. Is there a better way?